pp/wal/hashdex/scraper/encoding.h

```diff
@@ -0,0 +1,274 @@
+#pragma once
+
+#include <bit>
+#include <cmath>
+#include <cstdint>
+
+#include "bare_bones/bit.h"
+#include "primitives/sample.h"
+#include "prometheus/value.h"
+
+namespace PromPP::WAL::hashdex::scraper::encoding {
+enum class SampleValueType : uint8_t { kUint32 = 0b0000'0000, kDouble, kUint8, kUint16, kFloat, kZero, kNaN };
+
+[[nodiscard]] PROMPP_LAMBDA_INLINE inline SampleValueType value_type(const double val) noexcept {
+  if (std::isnan(val)) [[unlikely]] {
+    return SampleValueType::kNaN;
+  }
+  if (val == 0.0) [[unlikely]] {
+    return SampleValueType::kZero;
+  }
+
+  if (std::trunc(val) == val && val > 0.0 && val <= std::numeric_limits<uint32_t>::max()) [[likely]] {
+    const auto uval = static_cast<uint64_t>(val);
+    if (uval <= std::numeric_limits<uint8_t>::max()) {
+      return SampleValueType::kUint8;
+    }
+    if (uval <= std::numeric_limits<uint16_t>::max()) {
+      return SampleValueType::kUint16;
+    }
+    return SampleValueType::kUint32;
+  }
+
+  if (const auto f = static_cast<float>(val); static_cast<double>(f) == val) [[unlikely]] {
+    return SampleValueType::kFloat;
+  }
+
+  return SampleValueType::kDouble;
+}
+
+struct LayoutMarker {
+  uint8_t raw;
+
+  [[nodiscard]] PROMPP_LAMBDA_INLINE bool has_timestamp() const noexcept { return raw & 0b1000'0000; }
+
+  [[nodiscard]] PROMPP_LAMBDA_INLINE uint8_t count_size_in_bytes() const noexcept { return (raw >> 4) & 0b0000'0111; }
```

Comment:
count size in bytes is unclear name in my opinion. Maybe
```suggestion
  [[nodiscard]] PROMPP_LAMBDA_INLINE uint8_t size_bytes_count() const noexcept { return (raw >> 4) & 0b0000'0111; }
```
or
```suggestion
  [[nodiscard]] PROMPP_LAMBDA_INLINE uint8_t size_length_in_bytes() const noexcept { return (raw >> 4) & 0b0000'0111; }
```